In [1]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [6]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [7]:
sumar_nulos = (wh.isnull().sum())

In [30]:
sumar_nulos

Overall rank                     0
Country or region                0
Score                           19
GDP per capita                  12
Social support                  16
Healthy life expectancy         26
Freedom to make life choices    23
Generosity                      21
Perceptions of corruption       13
year                             0
dtype: int64

In [20]:
agrupados_por_anio = wh.groupby("year").count()

In [29]:
agrupados_por_anio

Overall rank  Country or region  Score  GDP per capita  Social support  \
year                                                                           
2018           156                156    151             147             141   
2019           156                156    142             153             155   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                      144                           144         139   
2019                      142                           145         152   

      Perceptions of corruption  
year                             
2018                        149  
2019                        150

In [27]:
list_years = pd.unique(wh["year"])

In [28]:
list_years

array([2018, 2019], dtype=int64)

In [9]:
listado_de_anios = {2018: {}, 2019:{}}

In [23]:
for pos in wh.columns[1 : -1]:
    listado_de_anios[2018][pos] = wh.loc[wh["year"] == 2018][pos].isnull().sum()/agrupados_por_anio[pos][2018]*100
    listado_de_anios[2019][pos] = wh.loc[wh["year"] == 2019][pos].isnull().sum()/agrupados_por_anio[pos][2019]*100

In [24]:
porcentajes_nulos = pd.DataFrame(listado_de_anios)

In [25]:
porcentajes_nulos

2018      2019
Country or region              0.000000  0.000000
Score                          3.311258  9.859155
GDP per capita                 6.122449  1.960784
Social support                10.638298  0.645161
Healthy life expectancy        8.333333  9.859155
Freedom to make life choices   8.333333  7.586207
Generosity                    12.230216  2.631579
Perceptions of corruption      4.697987  4.000000

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [32]:
datos_agrupados = wh.groupby("year")

In [34]:
datos_agrupados.describe()

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  142.0   

                ... Generosity        Perceptions of corruption            \
          mean  ...        75%    max                     count      mean   
year            ...                                                         
2018  5.354060  ...    0.23400  0.598                     149.0  0.112859   
2019  5.434028  ...    0.24825  0.566                     150.0  0.109053   

                                                     
           std  min      25%    50%      75%    max  
year                                                 
2018  0.098108  0.0  0.05000  0.082  0.13800  0.457  
2019  0.093279  0.0  0.04775  0.085  0.14025  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto



### resolucion ejercicio tres

In [36]:
promedio = wh.loc[wh["year"] == wh.fillna] = datos_agrupados.mean()

In [37]:
promedio

Overall rank     Score  GDP per capita  Social support  \
year                                                           
2018          78.5  5.354060        0.892748        1.205915   
2019          78.5  5.434028        0.900758        1.208265   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                 0.597576                      0.456514    0.181360   
2019                 0.722965                      0.395441    0.185092   

      Perceptions of corruption  
year                             
2018                   0.112859  
2019                   0.109053

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro